### ISSUE TITLE
Compiler error @ dev branch

### ISSUE THREAD

**Issue by da-phil** [Beep6581/RawTherapee#4085]
I just pulled the latest commits from dev branch and encountered the following compiler error:
``` c++
/home/phil/code/rawtherapee-git/rtengine/rtlensfun.cc: In static member function ‘static bool rtengine::LFDatabase::init(const Glib::ustring&)’:
/home/phil/code/rawtherapee-git/rtengine/rtlensfun.cc:270:92: error: ‘struct lfDatabase’ has no member named ‘LoadDirectory’
     bool ok = dbdir.empty() ? (instance_.data_->Load() ==  LF_NO_ERROR) : instance_.data_->LoadDirectory(dbdir.c_str());
```

Version:
```
-- 	Commit description:	5.2-171-g91fd578
-- 	Branch:			dev
-- 	Commit:			91fd578
-- 	Commit date:		2017-09-15
-- 	Commits since tag:	171
-- 	Commits since branch:	171
-- 	Version (unreliable):	5.2.171
```

**Comment by heckflosse** [comment]
@da-phil Which version of lensfun do you use? Stable or development?

**Comment by heckflosse** [comment]
maybe @agriggio also needs to chime in...

**Comment by heckflosse** [comment]
@da-phil Please post full rt verrsion information instead of crippled

**Comment by agriggio** [comment]
I'll take a look asap

**Comment by heckflosse** [comment]
@agriggio I get this when building on windows. My lensfun definitely has LoadDirectory() function.

```
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY - Failed
```

Edit: Same on my Manjaro using lensfun 0.3.2.0

**Comment by agriggio** [comment]
damn. it should still work however

**Comment by agriggio** [comment]
I'll see what goes wrong with the test

**Comment by heckflosse** [comment]
@agriggio 

according to [this](https://cmake.org/cmake/help/v3.0/module/CheckCXXSourceCompiles.html) the test also checks if linkage works. Maybe that's the reason the test goes wrong

**Comment by agriggio** [comment]
I've just pushed a fix for it. Works here, tested both with 0.3.2 (success) and 0.2.8 (failure, as expected). @heckflosse if you can test on windows and confirm it works, then we can close

**Comment by agriggio** [comment]
> according to this the test also checks if linkage works

Yes, that's basically the reason. But with the complication that CMake doesn't play 100% nicely with pkg-config, so the fix was more complicated than I thought.

In fact, this is almost always the issue with CMake -- everything is way more complicated than it should...  :angry:

**Comment by heckflosse** [comment]
I will test now

**Comment by heckflosse** [comment]
Tested on windows:

```
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY - Success
```

:+1:

**Comment by heckflosse** [comment]
Tested on Manjaro

```
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY
-- Performing Test LENSFUN_HAS_LOAD_DIRECTORY - Failed
```

CMakeError.log
```
Performing C++ SOURCE FILE Test LENSFUN_HAS_LOAD_DIRECTORY failed with the following output:
Change Dir: /home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp

Run Build Command:"/usr/bin/make" "cmTC_71695/fast"
/usr/bin/make -f CMakeFiles/cmTC_71695.dir/build.make CMakeFiles/cmTC_71695.dir/build
make[1]: Entering directory '/home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp'
Building CXX object CMakeFiles/cmTC_71695.dir/src.cxx.o
/usr/bin/c++   -I/usr/include/lensfun -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include  -std=c++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations -DLENSFUN_HAS_LOAD_DIRECTORY   -o CMakeFiles/cmTC_71695.dir/src.cxx.o -c /home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp/src.cxx
/home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp/src.cxx: In function ‘int main()’:
/home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp/src.cxx:5:10: warning: unused variable ‘b’ [-Wunused-variable]
     bool b = db->LoadDirectory(0);
          ^
Linking CXX executable cmTC_71695
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_71695.dir/link.txt --verbose=1
/usr/bin/c++   -std=c++11 -march=native -Werror=unused-label -fopenmp -Werror=unknown-pragmas -Wall -Wno-unused-result -Wno-deprecated-declarations -DLENSFUN_HAS_LOAD_DIRECTORY    -rdynamic CMakeFiles/cmTC_71695.dir/src.cxx.o  -o cmTC_71695 
CMakeFiles/cmTC_71695.dir/src.cxx.o: In function `main':
src.cxx:(.text+0x1d): undefined reference to `lfDatabase::LoadDirectory(char const*)'
collect2: error: ld returned 1 exit status
make[1]: *** [CMakeFiles/cmTC_71695.dir/build.make:98: cmTC_71695] Error 1
make[1]: Leaving directory '/home/ingo/repo-rt2/testbuild/CMakeFiles/CMakeTmp'
make: *** [Makefile:126: cmTC_71695/fast] Error 2

Source file was:
#include <lensfun.h>
int main()
{
    lfDatabase *db = 0;
    bool b = db->LoadDirectory(0);
    return 0;
}
```

**Comment by Beep6581** [comment]
`dev` compiles fine here in Sabayon: GCC-4.9.3, lensfun-0.3.2.

**Comment by agriggio** [comment]
@heckflosse did I mention how crazy CMake is? :smile: (I'm really amazed that it has become mainstream despite all its quirks...). Can you try this patch?

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -449,8 +449,12 @@
 include(CheckCXXSourceCompiles)
 set(CMAKE_REQUIRED_INCLUDES ${LENSFUN_INCLUDE_DIRS})
 set(CMAKE_REQUIRED_LIBRARIES)
+set(_lf_pth)
+if(LENSFUN_LIBRARY_DIRS)
+    set(_lf_pth NO_DEFAULT_PATH)
+endif()
 foreach(l ${LENSFUN_LIBRARIES})
-    find_library(_l ${l} PATHS ${LENSFUN_LIBRARY_DIRS} NO_DEFAULT_PATH)
+    find_library(_l ${l} PATHS ${LENSFUN_LIBRARY_DIRS} ${_lf_pth})
     set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} ${_l})
 endforeach()
 check_cxx_source_compiles(
```

**Comment by agriggio** [comment]
Or this one, which does the same but tries to be clearer on what is happening...

```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -450,7 +450,17 @@
 set(CMAKE_REQUIRED_INCLUDES ${LENSFUN_INCLUDE_DIRS})
 set(CMAKE_REQUIRED_LIBRARIES)
 foreach(l ${LENSFUN_LIBRARIES})
-    find_library(_l ${l} PATHS ${LENSFUN_LIBRARY_DIRS} NO_DEFAULT_PATH)
+    if(LENSFUN_LIBRARY_DIRS)
+        # the NO_DEFAULT_PATH is to make sure we find the lensfun version we
+        # want, and not the system's one (e.g. if we have a custom version
+        # installed in a non-standard location)
+        find_library(_l ${l} PATHS ${LENSFUN_LIBRARY_DIRS} NO_DEFAULT_PATH)
+    else()
+        # LENSFUN_LIBRARY_DIRS can be empty if lensfun is installed in the
+        # default path. In this case, adding NO_DEFAULT_PATH would make
+        # find_library fail...
+        find_library(_l ${l})
+    endif()
     set(CMAKE_REQUIRED_LIBRARIES ${CMAKE_REQUIRED_LIBRARIES} ${_l})
 endforeach()
 check_cxx_source_compiles(
```

**Comment by heckflosse** [comment]
@agriggio 
the last patch works fine on Manjaro and Windows

**Comment by agriggio** [comment]
closed via 33867a4bf22c9c8356e7980ed7bfb56fd5a8a82a
